cf-agent/package_module.c

```diff
@@ -0,0 +1,1611 @@
+                            const char *name, const char *ver, const char *arch)
+{
+    const char *version = ver;
+    /* Handle latest version in specific way for repo packages. 
+     * Please note that for file packages 'latest' version is not supported
+     * and check against that is made elsewhere. */
+    if (version && StringSafeEqual(version, "latest"))
+    {
+        version = NULL;
+    }
+    
+    /* Make sure cache is updated. */
+    if (ctx)
+    {
+        if (!UpdateSinglePackageModuleCache(ctx, module_wrapper,
+                                            UPDATE_TYPE_INSTALLED, false))
+        {
+            Log(LOG_LEVEL_ERR, "Can not update cache.");
+        }
+    }
+    
+    CF_DB *db_cached;
+    if (!OpenSubDB(&db_cached, dbid_packages_installed,
+                   module_wrapper->package_module->name))
+    {
+        Log(LOG_LEVEL_INFO, "Can not open cache database.");
+        return -1;
+    }
+    
+    char *key = NULL;
+    if (version && arch)
+    {
+        key = StringFormat("N<%s>V<%s>A<%s>", name, version, arch);
+    }
+    else if (version)
+    {
+        key = StringFormat("N<%s>V<%s>", name, version);
+    }
+    else if (arch)
+    {
+        key = StringFormat("N<%s>A<%s>", name, arch);
+    }
+    else
+    {
+         key = StringFormat("N<%s>", name);
+    }
+    
+    int is_in_cache = 0;
+    char buff[1];
+    
+    Log(LOG_LEVEL_DEBUG, "looking for key in cache: %s", key);
+    
+    if (ReadDB(db_cached, key, buff, 1))
+    {
+        /* Just make sure DB is not corrupted. */
+        if (buff[0] == '1')
+        {
+            is_in_cache = 1;
+        }
+        else
+        {
+            Log(LOG_LEVEL_INFO,
+                "Seem to have corrupted data in cache database");
+            is_in_cache = -1;
+        }
+    }
+    
+    Log(LOG_LEVEL_DEBUG,
+        "Looking for package %s in cache returned: %d", name, is_in_cache);
+    
+    CloseDB(db_cached);
+    
+    return is_in_cache;
+}
+
+void WritePackageDataToDB(CF_DB *db_installed,
+        const char *name, const char *ver, const char *arch,
+        UpdateType type)
+{
+    char package_key[strlen(name) + strlen(ver) +
+                     strlen(arch) + 11];
+    
+    xsnprintf(package_key, sizeof(package_key),
+              "N<%s>", name);
+    if ((type == UPDATE_TYPE_UPDATES || type == UPDATE_TYPE_LOCAL_UPDATES) && 
+            HasKeyDB(db_installed, package_key, strlen(package_key) + 1))
+    {
+        size_t val_size =
+                ValueSizeDB(db_installed, package_key, strlen(package_key));
+        char buff[val_size + strlen(arch) + strlen(ver) + 8];
+        
+        ReadDB(db_installed, package_key, buff, val_size);
+        xsnprintf(buff + val_size, sizeof(package_key), "V<%s>A<%s>\n", ver, arch);
+        WriteDB(db_installed, package_key, buff, strlen(buff));
+    }
+    else if (type == UPDATE_TYPE_UPDATES || type == UPDATE_TYPE_LOCAL_UPDATES)
```

Comment:
Repeated condition.  May be better to structure as:
```
if (type == UPDATE_TYPE_UPDATES || type == UPDATE_TYPE_LOCAL_UPDATES)
{
    if (HasKeyDB(db_installed, package_key, strlen(package_key) + 1)) {/* block above */}
    else {/* this block */}
}
else /* UPDATE_TYPE_INSTALLED */ {...}
```
or, if that last comment is to be trusted, even as
```
if (type == UPDATE_TYPE_INSTALLED) {/* the final else block below *}
else if (HasKeyDB(db_installed, package_key, strlen(package_key) + 1)) {/* the block above */}
else {/* the following block */}
```